jep/0000/README.adoc

```diff
@@ -0,0 +1,469 @@
+layer. Starting with the first boot, the client must frequently inform the
+server of its current versions, retrieve new versions, and apply updates. This
+document specifies those API interactions and expectations to allow this update
+lifecycle to be performed correctly.
+
+== Specification
+
+There are two services involved behind the scenes which are responsible for
+facilitating the client's update lifecycle: `updates` and `versions`.
+
+The role of the versions service is simply to maintain an audit trail
+for the versions of software that each client has installed locally. As clients
+go through various upgrade processes, they will periodically `POST`
+<<version-manifest>> records which are stored in the Evergreen hosted service
+layer's storage.
+
+The role of the update service is to use the `essentials.yaml` and retrieve the
+latest versions of software for a given client UUID and compute the appropriate
+update response to be sent to the client.
+
+.Bootstrapping Instance
+[source]
+----
+
+  update                     versions                    evergreen
+  service                    service                      client
+     |                           |                           |
+     |                           |     POST version manifest |
+     |                           |     from local hpis/core  |
+     |                           |<--------------------------o
+     |                           o-----------/ok/----------->|
+     |                           |                           |
+     |                     GET /update                       |
+     |<------------------------------------------------------o
+     |    GET /versions for uuid |                           |
+     o-------------------------->|                           |
+     |<-----/200 json payload/---o                           |
+     |                           |                           |
+     |            200 with computed JSON update manifest     |
+     |            indicating what files should               |
+     |            be downloaded                              |
+     o------------------------------------------------------>|
+     |                           |                           |
+----
+
+[[update-levels]]
+=== Update Levels
+
+In the backend datastore, updates must be considered immutable and sequential.
+These are referred to hereafter as "Update Levels". A single update record
+must have the following data structure:
+
+* `id`: Numerical sequential integer to identify the update
+* `commit`: a Git commit corresponding to a change of `essentials.yaml` in the
+  link:https://github.com/jenkins-infra/evergreen[jenkins-infra/evergreen]
+  repository
+* `channel`: String denoting the <<channel>> for the update.
+* `manifest`: JSON formatted representation of `essentials.yaml` for ease of
+  storage and query in PostgreSQL
+* `tainted`: Boolean to indicate whether this update should be considered
```

Comment:
I'm wondering if we should have a richer structure for _taintedness_. Is this really bad or good? I fear it might not be that simple. (I'm looking at instances that may hint a given UL is bad, when actually it's due to an _inessentials_ plugin that is installed specifically there)

Reply:
Honestly, I think this is something we'll have to figure out once Jenkins Essentials is in the field. We can add additional columns in the future without difficulty, but I don't want to spend too much time hunting for ghosts right now.